dhall/src/Dhall/Parser/Expression.hs

```diff
@@ -954,6 +954,26 @@ parsers embedded = Parsers {..}
 
             nonEmptyListLiteral <|> emptyListLiteral) <?> "literal"
 
+-- | Parse a single comment
+comment :: Parser Comment
+comment = LineComment <$> lineComment <|> BlockComment <$> blockComment <?> "comment"
+
+-- | Parse a multi comment (which contains at least one comment)
+multiComment :: Parser MultiComment
+multiComment = setCommentControl CommentIsNeeded $ -- Strict whitespace
+    MultiComment <$> Combinators.NonEmpty.sepEndBy1 comment whitespace
+
+-- | Parse a multi comment or whitespace
+commentOrWhitespace :: Parser (Maybe MultiComment)
+commentOrWhitespace = setCommentControl CommentIsNeeded $ -- Strict whitespace
+    do whitespace; optional multiComment
+
+-- | Parse a multi comment or non-empty whitespace
+commentOrNonEmptyWhitespace :: Parser (Maybe MultiComment)
+commentOrNonEmptyWhitespace = setCommentControl CommentIsNeeded $ -- Strict whitespace
+        try (Just <$ whitespace <*> multiComment)
```

Comment:
Generally you want to minimize use of `try`, both for performance reasons and better error messages.  What I'd suggest here to remove the dependence on `try` is something like:

```haskell
commentOrNonEmptyWhitespace = setCommentControl CommentIsNeeded $ do
    m <- optional whitespaceChunk

    case m of
        Nothing -> Just <$> multiComment
        Just () -> whitespace *> optional multiComment 
```

Reply:
Great, thanks for the explanation! :smile:

Reply:
You're welcome! 🙂